### PR TITLE
Handle max submissions limit in Canvas

### PR DESCRIPTION
**Pull request by marcospri**
Addressing some issues around canvas submissions:

https://hypothes-is.slack.com/archives/C2C2U40LW/p1677008765270819

From that thread and some debugging nothing seem to be completely broken around this feature but:

- We seem to get a lot of timeouts while taking with Canvas & this feature. This PR increases the timeout of that call.

https://github.com/hypothesis/lms/pull/5102

- There's an easy to reproduce issue where the LTI1.3 API honors the "max number of submissions" that can be configured in the UI https://github.com/hypothesis/lms/issues/4973

The PR won't offer any indication to the student that they hit that limit but it seems good enough to reduce the noise around this, see for a discussion about this:

https://hypothes-is.slack.com/archives/C4K6M7P5E/p1675419187872529

We'll need to document this in the user facing instructions as well

With this two changes I hope we can fix some of the timeouts and with less noise maybe discover some underlying issue.

### Submission attempts limit

To tests this run the server with:

`FEATURE_FLAG_SUBMIT_ON_ANNOTATION=true make dev`

 (looking into making this not an option as it has been the default for months)

You can check the configuration of these two assignments, the have "allowed  attempts" set to 1:

LTI1.3
https://hypothesis.instructure.com/courses/319/assignments/4700/edit?return_to=https%3A%2F%2Fhypothesis.instructure.com%2Fcourses%2F319%2Fassignments%2F4700

LTI1.1
https://hypothesis.instructure.com/courses/125/assignments/4691/edit?return_to=https%3A%2F%2Fhypothesis.instructure.com%2Fcourses%2F125%2Fassignments%2F4691
 

Now as `Hypothesis 101 Student` launch the both assignments:

#### LTI 1.1

The LTI1.1 version doesn't seem to handle this on the Canvas side:

- Launch https://hypothesis.instructure.com/courses/125/assignments/4691
make an annotation, refresh the page and make another, you don't get an error.

- Confirm the submission are made (as a teacher) here

https://hypothesis.instructure.com/courses/125/gradebook/speed_grader?assignment_id=4691&student_id=35

#### LTI 1.3

- Launch https://hypothesis.instructure.com/courses/319/assignments/4700 
make an annotation

In main you get an error dialog. In this branch you'll get... an line in the logs

```web (stderr)         | 2023-02-23 15:49:05,221 ERROR [lms.services.lti_grading._v13:77][MainThread] record_result: maximum number of allowed attempts```

### CODE REVIEW

File: lms/services/lti_grading/_v13.py [hunk 1]
```diff
@@ -56,13 +59,26 @@ def record_result(self, grading_id, score=None, pre_record_hook=None):
...
-            headers={"Content-Type": "application/vnd.ims.lis.v1.score+json"},
-        )
+        try:
+            return self._ltia_service.request(
+                "POST",
+                self._service_url(self.line_item_url, "/scores"),
+                scopes=self.LTIA_SCOPES,
+                json=payload,
+                headers={"Content-Type": "application/vnd.ims.lis.v1.score+json"},
+            ).json()
```
Comment (marcospri) [inline lms/services/lti_grading/_v13.py:69]: This is an extra change, now we return `.json()` not just the response itself.

That seems more inline with the rest of the service.

File: lms/services/lti_grading/_v13.py [hunk 2]
```diff
@@ -56,13 +59,26 @@ def record_result(self, grading_id, score=None, pre_record_hook=None):
...
+
+        except ExternalRequestError as err:
+            if (
+                err.status_code == 422
+                and "maximum number of allowed attempts has been reached"
+                in err.response.text
+            ):
+                LOG.error("record_result: maximum number of allowed attempts")
```
Comment (marcospri) [inline lms/services/lti_grading/_v13.py:77]: Easy to track in paper trail and gives an indication while testing locally.

File: lms/services/lti_grading/_v13.py [hunk 3]
```diff
@@ -56,13 +59,26 @@ def record_result(self, grading_id, score=None, pre_record_hook=None):
...
+                # We silently shallow this type of error
+                return {}
```
Comment (marcospri) [inline lms/services/lti_grading/_v13.py:79]: If we eventually want to show some modal to the student we can just make a change here.

I reckon an error modal might not be the right choice thought. This happens while creating the first annotation in a "session" so it's no clear for the student that they are in fact creating a submission.

They do get an error modal the moment they make that annotation but the annotation was made successful so I reckon this change is less confusing.

See https://hypothes-is.slack.com/archives/C4K6M7P5E/p1675419187872529 for further discussion.

File: lms/services/ltia_http.py [hunk 4]
```diff
@@ -58,6 +58,7 @@ def _get_access_token(self, scopes):
                 "client_assertion": signed_jwt,
                 "scope": " ".join(scopes),
             },
+            timeout=(20, 20),
```
Comment (marcospri) [inline lms/services/ltia_http.py:61]: This value seems aggressive enough (twice the default) but not crazy. We'll see how many more timeouts we see after.

File: lms/services/lti_grading/_v13.py (hunk 1 above)
Comment (seanh) [reply lms/services/lti_grading/_v13.py:69]: :+1:

**Review by seanh — APPROVED** [review]

**Pull request merged**